cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -424,18 +444,96 @@ func (c ControllerContext) IsControllerEnabled(controllerDescriptor *ControllerD
 	return genericcontrollermanager.IsControllerEnabled(controllerDescriptor.Name(), controllersDisabledByDefault, c.ComponentConfig.Generic.Controllers)
 }
 
-// InitFunc is used to launch a particular controller. It returns a controller
-// that can optionally implement other interfaces so that the controller manager
-// can support the requested features.
-// The returned controller may be nil, which will be considered an anonymous controller
-// that requests no additional features from the controller manager.
-// Any error returned will cause the controller process to `Fatal`
-// The bool indicates whether the controller was enabled.
-type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+// NewClientConfig is a shortcut for ClientBuilder.Config. It wraps the error with an additional message.
+func (c ControllerContext) NewClientConfig(name string) (*restclient.Config, error) {
+	config, err := c.ClientBuilder.Config(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client config for %q: %w", name, err)
+	}
+	return config, nil
+}
+
+// NewClient is a shortcut for ClientBuilder.Client. It wraps the error with an additional message.
+func (c ControllerContext) NewClient(name string) (kubernetes.Interface, error) {
+	client, err := c.ClientBuilder.Client(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for %q: %w", name, err)
+	}
+	return client, nil
+}
+
+// Runnable represents a loop implementing any piece of logic really.
+type Runnable interface {
+	// Run implements any kind of processing loop.
+	// When there is anything to be done, it blocks until the context is cancelled.
+	// Run must ensure all goroutines are terminated before returning.
+	Run(context.Context)
+}
+
+// Controller defines the base interface that all controller wrappers must implement.
+type Controller interface {
+	// Name returns the controller's canonical name.
+	Name() string
+
+	// Runnable is the main controller loop.
+	Runnable
+}
+
+// runnableFunc makes it possible to turn a function into a Runnable quickly.
+type runnableFunc func(context.Context)
+
+func (run runnableFunc) Run(ctx context.Context) {
+	run(ctx)
+}
+
+// runnables is a slice of Runnable objects that are to be run concurrently.
+type runnables []Runnable
+
+func (rx runnables) Run(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(len(rx))
+	for _, r := range rx {
+		go func() {
+			defer wg.Done()
+			r.Run(ctx)
+		}()
+	}
+	wg.Wait()
+}
+
+func runRunnables(ctx context.Context, rx ...Runnable) {
+	runnables(rx).Run(ctx)
+}
+
+// namedRunnable implements the Controller interface.
+type namedRunnable struct {
```

Comment:
Is this the only implementation of `Controller`?  Can we streamline this abstraction? I'm not sure I fully follow why both `Controller` and `namedRunnable` are needed.  What is missing if you just have a `type NamedRunnabel struct` ?

Reply:
No, some controllers, at least `garbageCollectorController`, implement the interface directly, because they also implement `controller.Debuggable` interface that needs to be accessible by embedding the original controller. So the idea is to use `namedRunnable` when simple wrapping is all that you need, but if you need to export any `controller.*` interface, you can do so as with the GC controller.